media_driver/agnostic/gen11/codec/hal/codechal_kernel_header_g11.h

```diff
@@ -160,7 +195,11 @@ static MOS_STATUS GetCommonKernelHeaderAndSizeG11(
     *((PCODECHAL_KERNEL_HEADER)krnHeader) = *currKrnHeader;
 
     PCODECHAL_KERNEL_HEADER invalidEntry;
+#ifdef _FULL_OPEN_SOURCE
     invalidEntry = &(kernelHeaderTable->weightedPrediction) + 1;
+#else
+    invalidEntry = &(kernelHeaderTable->dsConvertGenX0) + 1;
```

Comment:
Oh my God. That's really ugly!!! How about:
```
struct HmeDsScoreboardKernelHeaderG11 {
    int nKernelCount;
    union
    {
        struct
        {
            CODECHAL_KERNEL_HEADER hmeDownscaleGenX0;
            CODECHAL_KERNEL_HEADER hmeDownscaleGenX1;
            CODECHAL_KERNEL_HEADER hmeDownscaleGenX2;
            CODECHAL_KERNEL_HEADER hmeDownscaleGenX3;
            CODECHAL_KERNEL_HEADER hmeP;
            CODECHAL_KERNEL_HEADER hmeB;
            CODECHAL_KERNEL_HEADER hmeVdenc;
            CODECHAL_KERNEL_HEADER hmeHevcVdenc;
            CODECHAL_KERNEL_HEADER dsConvertGenX0;
#ifndef _FULL_OPEN_SOURCE
            CODECHAL_KERNEL_HEADER hmeDetectionGenX0;
            CODECHAL_KERNEL_HEADER initSwScoreboard;
            CODECHAL_KERNEL_HEADER intraDistortion;
            CODECHAL_KERNEL_HEADER dynamicScaling;
            CODECHAL_KERNEL_HEADER weightedPrediction;
#endif
            CODECHAL_KERNEL_HEADER invalidEntry;
        };
    };
};

invalidEntry = &kernelHeaderTable->invalidEntry;

```

Reply:
Looks good, but we cannot change the original kernel definition sequences, it will require current kernel binary/tools change, not only for Linux open source, also will impact Windows

Reply:
In that case we need to refactor for Windows as well. What exist in the code now is simply wrong and error prone. Actually, I don't quite understand what's the reason behind having pointer to the invalid entry?! Can you, please, explain more? In a way, what you have in a code now gives a question: do you have access violation hidden in the code? Why you store pointer to the address _after_ your structure if you can have sizeof(structure)?